File: src/Test/SubscriberForceExecuting.cs
﻿using SecretNest.MessageBus.Options;
using System;

namespace Test
{
    [TestClass]
    public class SubscriberForceExecuting
    {
        [TestMethod]
        public void TestMethod()
        {
            using var bus = new MessageBus();
            var subscriber1000Ticket = bus.RegisterSubscriber<int, int>("ArgumentMatching", Subscriber1000,
                new MessageBusSubscriberOptions<int, int>(sequence: 0, resultCheckingCallback: i => i > 0));
            var subscriber100Ticket = bus.RegisterSubscriber<int, int>("ArgumentMatching", Subscriber100,
                new MessageBusSubscriberOptions<int, int>(sequence: 2, resultCheckingCallback: i => i > 0));
            var subscriberForceTicket = bus.RegisterSubscriber<int>("ArgumentMatching", SubscriberForce,
                new MessageBusSubscriberOptions<int>(sequence: 1, isAlwaysExecution: true, isFinal: false));
       
            var publisherTicket = bus.RegisterPublisher<int, int>("ArgumentMatching", new MessageBusPublisherOptions<int, int>(defaultReturnValue: 5));

            _forceExecuted = false;
            var result1000 = publisherTicket.Executor.Execute(1500);
            Assert.IsTrue(_forceExecuted);
            _forceExecuted = false;
            var result100 = publisherTicket.Executor.Execute(300);
            Assert.IsTrue(_forceExecuted);
            _forceExecuted = false;
            var result5 = publisherTicket.Executor.Execute(80);
            Assert.IsTrue(_forceExecuted);

            bus.UnregisterPublisher(publisherTicket);
            bus.UnregisterSubscriber(subscriber1000Ticket);
            bus.UnregisterSubscriber(subscriber100Ticket);
            bus.UnregisterSubscriber(subscriberForceTicket);

            Assert.AreEqual(1000, result1000);
            Assert.AreEqual(100, result100);
            Assert.AreEqual(5, result5);
        }

        public int Subscriber1000(int argument)
        {
            if (argument > 1000)
                return 1000;
            else
                return 0;
        }

        public int Subscriber100(int argument)
        {
            if (argument > 100)
                return 100;
            else
                return 0;
        }

        private bool _forceExecuted;
        public void SubscriberForce(int argument)
        {
            _forceExecuted = true;
        }
    }
}

File: src/TestVB/SubscriberForceExecuting.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports SecretNest.MessageBus
Imports SecretNest.MessageBus.Options

<TestClass>
Public Class SubscriberForceExecuting
    <TestMethod>
    Sub TestMethod()
        Using bus As New MessageBus
            Dim subscriber1000Ticket = bus.RegisterSubscriber(Of Integer, Integer)("ArgumentMatching", AddressOf Subscriber1000, New MessageBusSubscriberOptions(Of Integer, Integer)(sequence:=0, resultCheckingCallback:=Function(i) i > 0))
            Dim subscriber100Ticket = bus.RegisterSubscriber(Of Integer, Integer)("ArgumentMatching", AddressOf Subscriber100, New MessageBusSubscriberOptions(Of Integer, Integer)(sequence:=2, resultCheckingCallback:=Function(i) i > 0))
            Dim subscriberForceTicket = bus.RegisterSubscriber(Of Integer)("ArgumentMatching", AddressOf SubscriberForce, New MessageBusSubscriberOptions(Of Integer)(sequence:=1, isAlwaysExecution:=True, isFinal:=False))

            Dim publisherTicket = bus.RegisterPublisher(Of Integer, Integer)("ArgumentMatching", New MessageBusPublisherOptions(Of Integer, Integer)(defaultReturnValue:=5))

            _forceExecuted = False
            Dim result1000 = publisherTicket.Executor.Execute(1500)
            Assert.IsTrue(_forceExecuted)
            _forceExecuted = False
            Dim result100 = publisherTicket.Executor.Execute(300)
            Assert.IsTrue(_forceExecuted)
            _forceExecuted = False
            Dim result5 = publisherTicket.Executor.Execute(80)
            Assert.IsTrue(_forceExecuted)

            bus.UnregisterPublisher(publisherTicket)
            bus.UnregisterSubscriber(subscriber1000Ticket)
            bus.UnregisterSubscriber(subscriber100Ticket)
            bus.UnregisterSubscriber(subscriberForceTicket)

            Assert.AreEqual(1000, result1000)
            Assert.AreEqual(100, result100)
            Assert.AreEqual(5, result5)
        End Using
    End Sub

    Public Function Subscriber1000(argument As Integer) As Integer
        If argument > 1000 Then
            Return 1000
        Else
            Return 0
        End If
    End Function

    Public Function Subscriber100(argument As Integer) As Integer
        If argument > 100 Then
            Return 100
        Else
            Return 0
        End If
    End Function

    Private _forceExecuted As Boolean

    Public Sub SubscriberForce(argument As Integer)
        _forceExecuted = True
    End Sub
End Class
